src/axom/core/Array.hpp

```diff
@@ -120,10 +209,12 @@ bool operator!=(const Array<T>& lhs, const Array<T>& rhs);
  *
  *
  * \tparam T the type of the values to hold.
+ * \tparam dim The dimension of the array.
+ * \tparam memory The memory space of the array.
```

Comment:
`memory` doesn't appear to be a `\tparam` (and shouldn't be one)

Reply:
The plan was to eventually move the memory space from a member variable to a template parameter - see the last bullet pt on #633 

Is this something you would be opposed to/apprehensive about?  The motivation for that design would be to leverage the type system such that a user would always know whether an `Array` is in host or device space, meaning that transfers would always be explicit and implemented via casting.

Reply:
I think that might be somewhat limiting, since Array's might then only reasonably be members of classes that are templated on the memory space.

I was thinking that the `Array_view` might be templated on the memory space, while the array itself could take an umpire allocator.

Seems like there should be a larger discussion on the pros/cons of each approach.

Reply:
I definitely prefer the idea of encoding the memory space in the array type directly. Serac is hitting a number of issues when trying to use mfem's containers which determine memory spaces at runtime. An example of a common issue is: imagine you want to write a function that adds the contents of 3 arrays:

```cpp
MyArray<double> add(MyArray<double> a, MyArray<double> b, MyArray<double> c) {
  // implementation unimportant
}
```

What should happen if the user passes in 3 arrays from different memory spaces?
1. treat it as a runtime error and exit
2. copy data to a common memory space and perform the addition
- if so, which space?
- if so, where should the output MyArray data live?

mfem does a mix of all of those options, making it incredibly hard to reason about which operations are allowed, and where calculations are taking place. In my mind, being able to write code that admits such ambiguities is a mistake. If the developer's intention is to require that the arrays share a common memory space, then you can enforce that at compile time through the type system

```cpp
MyArray<double, GPU> add(MyArray<double, GPU> a, MyArray<double, GPU> b, MyArray<double, GPU> c) {
  // implementation unimportant
}

int main() {
  MyArray<double, GPU> a, b;
  MyArray<double, CPU> c;
  auto sum = add(a, b, c); // compilation error!
}
```

If it is essential that you have a runtime memory space container, I would ask that you implement that _in addition_ to the more strongly-typed containers that include the memory space as a template argument.

Reply:
As someone who's spent more time than I'd like with some of this both in MFEM and elsewhere. I can definitely understand the desire for templated memory space within the type.  It can be tough to get right.

However, I don't believe long term that's a great solution when you've got more and more execution spaces needing to be supported in a code. I mean right now we have HOST, CUDA, HIP, and SYCL and for application codes especially those on the larger side I wouldn't say that having this sort of thing templated is a desirable trait.  I'd say more of them would probably prefer to just make use of a combination of the RAJA Suite (RAJA, Umpire, and CHAI) to deal with this memory execution space issue, since we know that technique works quite well even if it's not always as performant as a native solution.

One of the other issues downsides not as often talked about templating is also compilation slowdowns and an increase in the executable size.

It seems to me that the templated version should be an optional version if that's needed. I mean in theory it could just be a child class of a general ndarray implementation where that sort of thing is done at runtime.

Reply:
Could you explain in more detail what aspects of specifying the allocator at compile-time are undesirable to you @rcarson3?

I don't really see the number of memory spaces as a big issue. Supporting a new memory space for this container could be accomplished with a few of lines of code. The idea of making a memory allocator a template argument on a container is a very common practice (e.g. C++ standard library, thrust, hemi, Kokkos).

It is true that class templates with many parameters can slow down compile times when trying to exhaustively instantiate all the possible template combinations, but I don't see how that applies to this discussion-- the memory space will only ever be a single template parameter with at most a half dozen possibilities, and it's unlikely more than 2-3 of them will ever be simultaneously enabled for a given build.

To reiterate: I'm not saying that a container with runtime-determined memory space shouldn't exist, but that it would be nice if it wasn't the only option. It would be nice if there were a way to enforce at compile time that certain quantities are expected to belong to a specific memory space.

Reply:
@samuelpmishLLNL it might just be that I've attended too many different talks from large code groups and my own experience working with these sort of things in various large material libraries that my first thoughts are going with the runtime option first as it can reduce maintenance burden down the road if you make use of LLNL's various abstraction layers. Also, I think it might be worth thinking about how you want the template memory space to be defined. If you do it on the memory resource space than that might be too restrictive as `UVM, pinned, and CUDA` for example would all be compatible with one another, but the templated params would require a lot more duplicated code to cover the difference cases for a function. However, if you have the template based on the execution space than yes I would agree with you that you would just have either the CPU or GPU to deal with which probably isn't too bad.

I think we're both agreement that that templated versions aren't a bad idea :) I just think I have more of a preference for a runtime version initially.

Reply:
Thanks for the feedback @samuelpmishLLNL and @rcarson3.

We discussed this at an axom meeting yesterday and generally arrived at the same solution.
We didn't come to a resolution, but the consensus was that memory would be a policy on the array class.
The default would be to use a runtime choice (i.e. axom's memory API which uses umpire when avaiable).
But we would also like to support compile time types when desired.

In addition to the issues outlined above (compile times, and type equivalence), another major factor is how well the axom Array class would work with untemplated code. E.g. when a class that is not templated on memory type wants to use an axom array as a class member, we would not want to have specify the memory type within the array's type.

Reply:
Templating on the memory space doesn't need to be viral. If you know which machine you intend to run on, you can configure sensible defaults for the device (e.g. change default to HIP to target a machine with AMD hardware). One possible approach is something like

```cpp
struct CPU{};
struct CUDA{};
struct HIP{};

// pick an appropriate device for the machine you intend to run on
using default_device = CUDA;

template < typename T, int n >
using device_array = axom::Array<T, n, default_device >;

template < typename T, int n >
using host_array = axom::Array<T, n, CPU >;

struct Simulation {

  // keep performance-critical data on the device
  device_array < double, 2 > displacements;
  device_array < double, 2 > temperature;
  device_array < material_info_t, 2 > material_state;
  ...

  // keep other quantities on the host (CPU, typically)
  host_array < std::string > datafiles;
  ...

};
```

This same struct definition is usable on computers with different accelerator devices (by reconfiguring the project + conditional compilation), does not instantiate any unnecessary templates (don't pay for what you don't use), and is immediately clear about where the data lives and who owns it.